Template Sidebar: Fix list of current template part areas to show nested template parts

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes #47213 

Ensure that the Template tab in the sidebar reflects all template parts in the template, irrespective of their level of nesting. This resolves a bug where nested header or footer template parts were not visible in the list.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Nesting template parts is useful to do, particularly for sticky headers, so we need to make sure nested template parts are still accessible via the Template navigation in the sidebar.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

The approach in this PR does two things:

* Iterate over all nested blocks in the template to locate the template parts.
* Optimise the selector so that it uses memoized calculation.

Overall this reduces the number of times that the processing of the template's blocks occurs. In my local environment, this reduces the processing from around 126 times with the list view open, to 3.

Detailed changes:

* Borrow `flattenBlocks` logic from the block editor reducer [here](https://github.com/WordPress/gutenberg/blob/7e7e78cb1ea1585811650416176779259f94430b/packages/block-editor/src/store/reducer.js#L80), but simplify the logic slightly.
* Introduce `memize` into the `edit-site` package (it is also used in many other packages).
* Move filtering of template blocks logic out of `getCurrentTemplateTemplateParts` and into a separate utility `getFilteredTemplatePartBlocks` function.
* Use a memoized version of that function `memoizedGetFilteredTemplatePartBlocks` in `getCurrentTemplateTemplateParts` selector.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. See if you can reproduce #47213 by wrapping a header block in a group block, and see that it's no longer visible in the Template tab in the site editor, without this PR
2. With this PR applied, if you nest a Group block, it should still be available in the template tab in the right hand sidebar

## Screenshots or screencast <!-- if applicable -->

| Before (header not visible in sidebar) | After (header visible in sidebar) |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/213096329-2470a1ec-610b-4092-b1ad-ec78763e2d5f.png) | ![image](https://user-images.githubusercontent.com/14988353/213096346-8a39be9a-610d-4eb7-8951-77d150cbc3d1.png) |

andrewserong committed:
Try: Update list of current template part areas to recursively check for template parts

andrewserong committed:
Simplify logic, add memoized call

andrewserong commented:
Hi folks! I've pinged a few of you as reviewers, as this PR attempts to improve the performance of a selector at the same time as fixing a bug. I don't think I've added a dependency to a package before, so just wanted to double-check that I'm not doing anything strange here, or if I've missed any potential issues with using `memize` in the `edit-site` package. For context, I noticed that the processing in `getCurrentTemplateTemplateParts` was happening quite frequently, and since this PR adds more complexity to that processing, I thought it'd probably benefit from some memoization.

youknowriad commented on packages/edit-site/src/store/selectors.js:
```diff
@@ -247,6 +248,61 @@ export function isSaveViewOpened( state ) {
+
+	return flattenBlocks( templateBlocks )
+		.filter( ( block ) => isTemplatePart( block ) )
+		.map( ( block ) => {
```
Here we're looping four times: inside flatter, the filter, the map and another filter. If we want maximum performance, we could copy the logic of "flattenBlocks" here and add the checks and mapping within the "while" loop. 

That said, the improvement could be too small to worth it in terms of code clarity, so really your call here, just wanted to comment it on that.

tyxla replied:
We could as well combine the additional filtering and mapping with `flattenBlocks()`, so it all would happen in that `while` block - `continue`-ing if the block is not a template block; mapping and again `continue`-ing if there's no `templatePart`. That would just be a single loop instead of several ones.

youknowriad commented on packages/edit-site/src/store/selectors.js:
```diff
@@ -268,32 +324,11 @@ export const getCurrentTemplateTemplateParts = createRegistrySelector(
+		return memoizedGetFilteredTemplatePartBlocks(
+			template.blocks || [],
+			templateParts
+		);
```
I'm guessing it's not possible to use createSelector with createRegistrySelector and that's why you had to rely on "memize" directly?

andrewserong replied:
Yes, that appears to be the case, unless I've missed something!

youknowriad reviewed (commented):
Feels like this selector might benefit from a small unit test with a nesting case (not sure it's possible to test easily though as it touches several stores)?

tyxla commented on packages/edit-site/src/store/selectors.js:
```diff
@@ -247,6 +248,61 @@ export function isSaveViewOpened( state ) {
+	return result;
+}
+
+function getFilteredTemplatePartBlocks( templateBlocks = [], templateParts ) {
```
When referencing an empty array, we could declare an `EMPTY_ARRAY` constant and reference it here. IIRC, `memize()` is doing a strict comparison for each argument, and an empty array will be a different object every time, thus ignoring the previously cached value.

andrewserong replied:
Good catch! I've switched over to use `EMPTY_ARRAY`.

tyxla commented on packages/edit-site/src/store/selectors.js:
```diff
@@ -268,32 +324,11 @@ export const getCurrentTemplateTemplateParts = createRegistrySelector(
-			.filter( ( { templatePart } ) => !! templatePart );
+
+		return memoizedGetFilteredTemplatePartBlocks(
+			template.blocks || [],
```
Do we really need the `|| []` if we have a default value for the first argument of `getFilteredTemplatePartBlocks()`. OTOH, if we specify the empty array here, why do we need the default value above? One of those seems redundant.

andrewserong replied:
Good point, I've switched over to using the `EMPTY_ARRAY` approach you suggested in the other comment.

andrewserong committed:
Move logic to a single function to reduce multiple loops, move to utils, add a test

andrewserong commented:
Thanks for the feedback @youknowriad and @tyxla, all good notes! 🙇 I've done the following:

* Combined the logic into a single function to reduce the number of times we need to loop over the blocks.
* Within the loop, use the logic to check if the block is a template part, and perform the template part lookup. Only add to the result if there's a match.
* Moved the function to a `utils` file so that we can export it so that it can be tested.
* Added a test to confirm that it properly iterates over nested items, and that those nested items preserve a depth-first order.

I _think_ this now gives us an adequate level of testing coverage for the change in this PR in that the selector is now only responsible for gathering the blocks from the template and the available template parts and then passing them along to `memoizedGetFilteredTemplatePartBlocks`.

Please let me know if you'd like to see other changes or if I've missed anything!

youknowriad commented on packages/edit-site/src/store/selectors.js:
```diff
@@ -10,10 +10,14 @@ import { store as coreDataStore } from '@wordpress/core-data';
+/**
+ * Internal dependencies
+ */
+import { memoizedGetFilteredTemplatePartBlocks } from './utils';
```
I guess we're forced to use another file because all exported functions here are selectors?

tyxla commented on packages/edit-site/src/store/utils.js:
```diff
@@ -0,0 +1,70 @@
+ * @param {?Array} templateParts Available template parts.
+ * @return {Array} An array of template parts and their blocks.
+ */
+export function getFilteredTemplatePartBlocks(
```
I think it can be confusing if we export both the non-memoized and the memoized versions. 

Could we just export `memoizedGetFilteredTemplatePartBlocks` and ideally, export it as `getFilteredTemplatePartBlocks` to avoid that confusion?

I see we're exposing the non-memoized function for unit testing purposes, but it should be fine to test the memoized function and just add a `memoizedGetFilteredTemplatePartBlocks.clear()` before or after every test, right?

andrewserong replied:
Looks like it works pretty well exporting the memoized version as `getFilteredTemplatePartBlocks`, and neatens things up a fair bit. Good idea 👍

tyxla commented on packages/edit-site/src/store/utils.js:
```diff
@@ -0,0 +1,70 @@
+	return result;
+}
+
+export const memoizedGetFilteredTemplatePartBlocks = memoize(
```
I think it won't hurt to have a test that verifies that caching works as expected.

andrewserong replied:
I've added a test — since we're no longer exporting the uncached version, I wound up needing to use the `getCache()` method that gets added by `memize` for the test — so it does wind up tying the test to `memize` a bit, but I think it's better to include the test than not at all.

tyxla reviewed (commented):
This looks great, thank you @andrewserong!

I've left a few more suggestions regarding the public API we're exposing and unit testing - let me know what you think.

andrewserong committed:
Add test to cover memoization, only export memoized function

andrewserong commented:
Thanks for the follow-up feedback @tyxla (and for the review, Riad!)! I've adjusted it so that we're only exporting the memoized version and added a test. Let me know if you think there's any other follow-ups to make there, otherwise I'll merge this one in tomorrow.

tyxla reviewed (approved):
Looks & works great, thanks @andrewserong 🚀

andrewserong commented:
Excellent, thanks again for the help in optimising this one! 🙇
